streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java

```diff
@@ -59,4 +89,55 @@ public SourceNode source(String topic) {
         return stateStoreSuppliers;
     }
 
+
+    private String childrenToString(List<ProcessorNode<?, ?>> children) {
+        if (children == null || children.isEmpty()) {
+            return null;
+        }
+
+        StringBuilder sb = new StringBuilder("children [");
+        for (ProcessorNode child : children) {
+            sb.append(child.name() + ",");
+        }
+        sb.setLength(sb.length() - 1);
+        sb.append("]\n");
+
+        // recursively print children
+        for (ProcessorNode child : children) {
+            sb.append("\t\t" + child.name() + ": ");
+            if (sinkNameToTopic.containsKey(child.name())) {
+                sb.append("topics:" + sinkNameToTopic.get(child.name()) + " ");
+            }
+            if (child.stateStores != null && !child.stateStores.isEmpty()) {
+                sb.append("stateStores [");
+                for (String store : (Set<String>) child.stateStores) {
+                    sb.append(store + ",");
+                }
+                sb.setLength(sb.length() - 1);
+                sb.append("] ");
+            }
+            if (!sinkNameToTopic.containsKey(child.name())) {
+                sb.append(childrenToString(child.children()));
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Produces a string representation contain useful information this topology.
+     * This is useful in debugging scenarios.
+     * @return A string representation of this instance.
+     */
+    public String toString() {
+        StringBuilder sb = new StringBuilder("\t\tProcessorTopology:\n");
+
+        // start from sources
+        for (String topic : sourceByTopics.keySet()) {
+            SourceNode source = sourceByTopics.get(topic);
+            sb.append("\t\t" + source.name() + ": " + "topics: " + sourceNameToTopic.get(source.name()) + ", ");
```

Comment:
In addition to the previous comment, some source nodes may not be mapped to a list of topics but regex pattern as well.